Arm/Memory/SeparateAutomation.lean

```diff
@@ -191,81 +191,119 @@ def getConfig : SimpMemM SimpMemConfig := do
 /-- info: state_value (fld : StateField) : Type -/
 #guard_msgs in #check state_value
 
+
+def SimpMemM.findOverlappingReadHypAux (hyps : Array Memory.Hypothesis) (er : ReadBytesExpr)  (hReadEq : ReadBytesEqProof) :
+    SimpMemM <| Option (MemSubsetProof { sa := er.span, sb := hReadEq.read.span }) := do
+  withTraceNode m!"{processingEmoji} ... ⊆ {hReadEq.read.span} ? " do
+    -- the read we are analyzing should be a subset of the hypothesis
+    let subset := (MemSubsetProp.mk er.span hReadEq.read.span)
+    let some hSubsetProof ← proveWithOmega? subset (← getBvToNatSimpCtx) (← getBvToNatSimprocs)  hyps
+      | return none
+    return some (hSubsetProof)
+
+def SimpMemM.findOverlappingReadHyp (hyps : Array Memory.Hypothesis) (er : ReadBytesExpr) :
+    SimpMemM <| Option (Σ (hread : ReadBytesEqProof),  MemSubsetProof { sa := er.span, sb := hread.read.span }) := do
+  for hyp in hyps do
+    let Hypothesis.read_eq hReadEq := hyp
+      | continue
+    let some subsetProof ← SimpMemM.findOverlappingReadHypAux hyps er hReadEq
+      | continue
+    return some ⟨hReadEq, subsetProof⟩
+  return none
+
+
+def Expr.kindStr (e : Expr) : String :=
```

Comment:
Is this different from `Lean.Expr.ctorName`?

Reply:
Very nice, I didn't notice this API. Switching to it now!